json11.cpp

```diff
@@ -571,182 +705,463 @@ struct JsonParser final {
 
         if (ch == '-' || (ch >= '0' && ch <= '9')) {
             i--;
-            return parse_number();
+            return set_state(VALUE_NUMBER);
         }
 
         if (ch == 't')
-            return expect("true", true);
+            return set_state(VALUE_TRUE);
 
         if (ch == 'f')
-            return expect("false", false);
+            return set_state(VALUE_FALSE);
 
         if (ch == 'n')
-            return expect("null", Json());
+            return set_state(VALUE_NULL);
 
         if (ch == '"')
-            return parse_string();
+            return set_state(VALUE_STRING);
 
-        if (ch == '{') {
-            map<string, Json> data;
-            ch = get_next_token();
-            if (ch == '}')
-                return data;
+        if (ch == '{')
+            return set_state(VALUE_OBJECT);
 
-            while (1) {
-                if (ch != '"')
-                    return fail("expected '\"' in object, got " + esc(ch));
+        if (ch == '[')
+            return set_state(VALUE_ARRAY);
 
-                string key = parse_string();
-                if (failed)
-                    return Json();
-
-                ch = get_next_token();
-                if (ch != ':')
-                    return fail("expected ':' in object, got " + esc(ch));
-
-                data[std::move(key)] = parse_json(depth + 1);
-                if (failed)
-                    return Json();
-
-                ch = get_next_token();
-                if (ch == '}')
-                    break;
-                if (ch != ',')
-                    return fail("expected ',' in object, got " + esc(ch));
+        return fail("expected value, got " + esc(ch));
+    }
 
-                ch = get_next_token();
+    void consume(const std::string &in = std::string()) {
+        need_data = false;
+        str += in;
+
+        /* try to parse as much as possible */
+        while (!states.empty()) {
+            switch (states.top()) {
+            case EXPECT_VALUE:
+                parse_json();
+                break;
+            case VALUE_OBJECT:
+            case OBJECT_KEY_OR_END:
+            case OBJECT_COMMA_OR_END:
+            case OBJECT_KEY:
+            case OBJECT_COLON:
+            case OBJECT_VALUE:
+                parse_object();
+                break;
+            case VALUE_ARRAY:
+            case ARRAY_VALUE_OR_END:
+            case ARRAY_COMMA_OR_END:
+            case ARRAY_VALUE:
+                parse_array();
+                break;
+            case VALUE_STRING:
+                parse_string();
+                break;
+            case VALUE_NUMBER:
+                parse_number();
+                break;
+            case VALUE_TRUE:
+                parse_true();
+                break;
+            case VALUE_FALSE:
+                parse_false();
+                break;
+            case VALUE_NULL:
+                parse_null();
+                break;
+            case VALUE_COMMENT:
```

Comment:
A comment isn't really a value, which makes this state name confusing.  It's hard for me to tell from the code if you actually have limitations on where comments can appear (e.g. only where a JSON value would go) or if it's just the name which is odd.